### DISCUSSION TITLE
Coral TPU worth it in small environment?

### DISCUSSION THREAD

**Discussion by jokakilla** [blakeblackshear/frigate#3963] [Ask A Question] [answered]
I only have two cameras running with 720p for the detection and 2688x1520 for recording.
I'm pretty impressed by the object detection and performance. On my Ryzen 5600G with GPU cpu usage is pretty low (especially compared to other solutions I tried before like AgentDVR along with DeepStack or MotionEye).

All the documentation and logs are recommending using Coral TPU and remind me that using the CPU is not a good idea.
As Coral USB TPU are currently hardly available and expensive I'm wondering if there is really a benefit in small environments like mine.

What can I expect from it? Probably even lower cpu usage but would the object detection really benefit? I'm already pretty happy with it.
Any experience?

**Comment by NickM-27** [comment, marked as answer]
Yes, it is still recommended. CPU usage is only part of the picture. The other part is latency / inference time.

Not sure what the inference time for your CPU is but in general it is ~~80-120 milliseconds for high end CPUs (from what I've seen anyway). 

Meanwhile, a USB coral is ~10 milliseconds and in my case my a key dual coral via PCIe adapter is 5 ms. Frigate often runs many detections per frame especially when there are multiple objects in the frame.

If you assumed an average of 3 detections per frame and your 2 cameras are both set to 5 frames per second that equates to (3 * 5) * 2 = 30 inferences per second. With an inference speed of 100 ms your CPU can do 10 inferences in a second which means you will have skipped frames. 

Naturally this is an example to illustrate and it doesn't always play out that way continually. But the point is there's still a benefit.

Another benefit is just the faster response time as there's less processing delay.

**Comment by blakeblackshear** [comment]
It will be a tiny investment for the benefit when prices come back down.

**Comment by jokakilla** [comment]
Thanks for your replies.
After reading your response I've had another look for the USB TPU and ordered one (in stock) for "only" 80€. Last time I looked for one the cheapest was 140€.

**Comment by jokakilla** [comment]
It arrived and installation was easy peasy. Inference is now down to 10 :)
